File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.java
import java.util.*;

class Solution {
    public int solution(int[][] points, int[][] routes) {
        int answer = 0;

        // 로봇 초기 위치 설정
        List<Map<Integer, Loc>> moves = new ArrayList<>();
        int maxTime = 0;
        for (int[] route : routes) {
            int time = 0;
            Map<Integer, Loc> map = new HashMap<>();
            map.put(0, new Loc(points[route[0]-1][0], points[route[0]-1][1]));
            for(int j = 1; j<route.length; j++) {
                // 시작위치
                int startX = points[route[j-1]-1][0];
                int startY = points[route[j-1]-1][1];
                // 도착위치
                int destX = points[route[j]-1][0];
                int destY = points[route[j]-1][1];

                int currentX = startX;
                int currentY = startY;

                if(destX > startX) {
                    for (int i = 0; i < destX-startX; i++) {
                        map.put(++time, new Loc(++currentX, currentY));
                    }
                }else if(destX < startX) {
                    for (int i = 0; i < startX-destX; i++) {
                        map.put(++time, new Loc(--currentX, currentY));
                    }
                }

                if(destY > startY) {
                    for (int i = 0; i < destY-startY; i++) {
                        map.put(++time, new Loc(currentX, ++currentY));
                    }
                }
                else if(destY < startY) {
                    for (int i = 0; i < startY-destY; i++) {
                        map.put(++time, new Loc(currentX, --currentY));
                    }
                }
            }
            maxTime = Math.max(maxTime, time);
            moves.add(map);
        }

        // 같은 시간에 같은 좌표 체크
        for (int time = 0; time <= maxTime; time++) {
            Map<Loc, Integer> times = new HashMap<>();
            for (Map<Integer, Loc> map : moves) {
                Loc loc = map.get(time);
                if( loc != null) {
                    times.put(map.get(time), times.getOrDefault(map.get(time), 0) + 1);
                }
            }
            for (Loc loc : times.keySet()) {
                if (times.get(loc) >1) {
                    answer ++;
                }
            }
        }
        return answer;
    }
//


    // Loc 클래스에 equals와 hashCode를 정의
    static class Loc {
        int x, y;

        public Loc(int x, int y) {
            this.x = x;
            this.y = y;
        }

        public int getX() {
            return x;
        }

        public int getY() {
            return y;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Loc loc = (Loc) o;
            return x == loc.x && y == loc.y;
        }

        @Override
        public int hashCode() {
            return Objects.hash(x, y);
        }
    }
}

File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.py
from collections import defaultdict, deque
def solution(points, routes):
    bol_d = defaultdict(int)
    answer = 0
    def go(sp,ep, t):
        start_y = points[sp-1][0] - 1
        start_x = points[sp-1][1] - 1
        end_y   = points[ep-1][0] - 1
        end_x   = points[ep-1][1] - 1
        # print((start_y, start_x), (end_y, end_x))
        # bol_d[(t, start_y, start_x)] += 1    
        while start_y != end_y:
            if start_y > end_y:
                start_y -= 1
            else:
                start_y += 1
            t += 1
            # print(t, start_y, start_x)
            bol_d[(t, start_y, start_x)] += 1

        while start_x != end_x:
            if start_x > end_x:
                start_x -=1
            else:
                start_x += 1
            t +=1
            # print(t, start_y, start_x)
            bol_d[(t, start_y, start_x)] += 1

        # print((sp, ep), bol_d)
        return t

    # for j in points:
    #     bol_d[(0, j[0]-1, j[1]-1)] += 1
    for i in range(len(routes)):
        rout = deque(routes[i])
        time = 0
        st = rout.popleft()
        while rout:
            ed = rout.popleft()
            bol_d[(time, points[st-1][0]-1, points[st-1][1]-1)] += 1
            time = go(st,ed, time)

            if rout:
                st = ed
                bol_d[(time, points[st-1][0]-1, points[st-1][1]-1)] -= 1
                # rout.appendleft(ed)
    # print(bol_d)
    for i in bol_d:
        if bol_d[i] > 1:
            answer += 1
    return answer